topology/probes/libvirt/libvirt.go

```diff
@@ -0,0 +1,267 @@
+	if err := libvirtgo.EventRegisterDefaultImpl(); err != nil {
+		logging.GetLogger().Errorf("libvirt event handler:  %s", err)
+	}
+
+	go func() {
+		for {
+			if err := libvirtgo.EventRunDefaultImpl(); err != nil {
+				logging.GetLogger().Errorf("libvirt poll loop problem: %s", err)
+			}
+		}
+	}()
+
+	conn, err := libvirtgo.NewConnectReadOnly(probe.uri)
+	if err != nil {
+		logging.GetLogger().Errorf("Failed to create libvirt connect")
+		return
+	}
+	probe.conn = conn
+	callback := func(
+		c *libvirtgo.Connect, d *libvirtgo.Domain,
+		event *libvirtgo.DomainEventLifecycle,
+	) {
+		if event.Event == libvirtgo.DOMAIN_EVENT_STARTED {
+			var interfaces []*Interface
+			probe.queryDomain(d, &interfaces, "")
+			probe.registerInterfaces(interfaces)
+		}
+	}
+	probe.cidLifecycle, err = conn.DomainEventLifecycleRegister(nil, callback)
+	if err != nil {
+		logging.GetLogger().Errorf(
+			"Could not register the lifecycle event handler %s", err)
+	}
+	callbackDeviceAdded := func(
+		c *libvirtgo.Connect, d *libvirtgo.Domain,
+		event *libvirtgo.DomainEventDeviceAdded,
+	) {
+		var interfaces []*Interface
+		probe.queryDomain(d, &interfaces, event.DevAlias)
+		probe.registerInterfaces(interfaces) // 0 or 1 device changed.
+	}
+	probe.cidDevAdded, err = conn.DomainEventDeviceAddedRegister(nil, callbackDeviceAdded)
+	if err != nil {
+		logging.GetLogger().Errorf(
+			"Could not register the device added event handler %s", err)
+	}
+	domains, err := probe.conn.ListAllDomains(0)
+	if err != nil {
+		logging.GetLogger().Error(err)
+		return
+	}
+	var interfaces []*Interface
+	for _, domain := range domains {
+		probe.queryDomain(&domain, &interfaces, "")
+	}
+	probe.registerInterfaces(interfaces)
+}
+
+// Stop stops the probe
+func (probe *Probe) Stop() {
+	probe.graph.RemoveEventListener(probe)
+	if probe.cidLifecycle != -1 {
+		probe.conn.DomainEventDeregister(probe.cidLifecycle)
+		probe.conn.DomainEventDeregister(probe.cidDevAdded)
+	}
+	probe.conn.Close()
+}
+
+// NewProbe creates a libvirt topology probe
+func NewProbe(g *graph.Graph, uri string) *Probe {
+	probe := &Probe{
+		graph:        g,
+		interfaceMap: make(map[string]*Interface),
+		cidLifecycle: -1,
+		uri:          uri,
+	}
+	g.AddEventListener(probe)
+	return probe
+}
+
+// NewProbeFromConfig initializes the probe
+func NewProbeFromConfig(g *graph.Graph) (*Probe, error) {
+	uri := config.GetString("agent.topology.libvirt.url")
+	probe := NewProbe(g, uri)
+	return probe, nil
+}
+
+// OnNodeUpdated event
+func (probe *Probe) OnNodeUpdated(n *graph.Node) {
+}
+
+// OnNodeAdded event
+func (probe *Probe) OnNodeAdded(n *graph.Node) {
+	if typ, _ := n.GetFieldString("Type"); typ != "tun" {
+		return
+	}
+	name, _ := n.GetFieldString("Name")
+	logging.GetLogger().Debugf("Tun interface %s", name)
+	if itf, ok := probe.interfaceMap[name]; ok {
+		probe.enrichNode(n, itf)
+	}
+}
+
+// OnNodeDeleted event
+func (probe *Probe) OnNodeDeleted(n *graph.Node) {
```

Comment:
I think we don't need this code. Basically, the interfaces will be removed when libvirt machine stops and then these interfaces will be added back when it will be started, so better to rely on handling of events from libvirt like it is done here.
https://github.com/skydive-project/skydive/pull/1424/files#diff-9934bc7c03f53af658e6552f294f1b7cR107